### PR TITLE
DATAGRAPH-1024 - Introduce reactive repository API.

### PR DESCRIPTION
**Pull request by LilitGabrielyan**
Add new annotation for reactive repositories detection.
New Reactive Repository.
Unit tests for all cases.

- [ ] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/master/CONTRIBUTING.adoc).
- [ ] There is a ticket in the bug tracker for the project in our [JIRA](https://jira.spring.io/browse/DATAGRAPH).
- [ ] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/master/etc/ide) and have them applied to your changes. Don’t submit any formatting related changes.
- [ ] You submit test cases (unit or integration tests) that back your changes.
- [ ] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).

### CODE REVIEW

File: spring-data-neo4j/src/main/java/org/springframework/data/neo4j/repository/query/ReactiveGraphQueryExecution.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+			}
+			Iterator<?> iterator = result.iterator();
+			if (!iterator.hasNext()) {
+				return null;
+			}
+			Object ret = iterator.next();
+			if (iterator.hasNext()) {
+				throw new IncorrectResultSizeDataAccessException("Incorrect result size: expected at most 1", 1);
+			}
+			return Mono.just(ret);
```
Comment (mp911de) [inline spring-data-neo4j/src/main/java/org/springframework/data/neo4j/repository/query/ReactiveGraphQueryExecution.java]: These are blocking calls hidden behind an API returning reactive types.

File: spring-data-neo4j/src/main/java/org/springframework/data/neo4j/repository/query/ReactiveGraphQueryExecution.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+	@RequiredArgsConstructor
+	final class CollectionExecution implements ReactiveGraphQueryExecution {
+
+		private final Session session;
+		private final ReactiveGraphParameterAccessor accessor;
+
+		@Override
+		public Object execute(Query query, Class<?> type) {
+			if (query.isFilterQuery()) {
```
Comment (mp911de) [inline spring-data-neo4j/src/main/java/org/springframework/data/neo4j/repository/query/ReactiveGraphQueryExecution.java]: These are blocking calls hidden behind an API returning reactive types.

File: spring-data-neo4j/src/main/java/org/springframework/data/neo4j/repository/query/ReactiveGraphQueryExecution.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+	@RequiredArgsConstructor
+	final class QueryResultExecution implements ReactiveGraphQueryExecution {
+
+		private final Session session;
+		private final ReactiveGraphParameterAccessor accessor;
+
+		@Override
+		public Object execute(Query query, Class<?> type) {
+			return Mono.just(session.query(query.getCypherQuery(accessor.getSort()), query.getParameters()));
```
Comment (mp911de) [inline spring-data-neo4j/src/main/java/org/springframework/data/neo4j/repository/query/ReactiveGraphQueryExecution.java]: This is a blocking call hidden behind an API returning reactive types.

File: spring-data-neo4j/src/main/java/org/springframework/data/neo4j/repository/query/ReactiveGraphQueryExecution.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+	}
+
+	@RequiredArgsConstructor
+	final class CountByExecution implements ReactiveGraphQueryExecution {
+
+		private final Session session;
+
+		@Override
+		public Object execute(Query query, Class<?> type) {
+			return Mono.just(session.count(type, query.getFilters()));
```
Comment (mp911de) [inline spring-data-neo4j/src/main/java/org/springframework/data/neo4j/repository/query/ReactiveGraphQueryExecution.java]: This is a blocking call hidden behind an API returning reactive types.

File: spring-data-neo4j/src/main/java/org/springframework/data/neo4j/repository/query/ReactiveGraphQueryExecution.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+	@RequiredArgsConstructor
+	final class DeleteByExecution implements ReactiveGraphQueryExecution {
+
+		private final Session session;
+		private final GraphQueryMethod graphQueryMethod;
+
+		@Override
+		public Object execute(Query query, Class<?> type) {
+			Class<?> returnType = graphQueryMethod.getReturnedObjectType();
```
Comment (mp911de) [inline spring-data-neo4j/src/main/java/org/springframework/data/neo4j/repository/query/ReactiveGraphQueryExecution.java]: This is a blocking call hidden behind an API returning reactive types.

**Comment by nmervaillie** [comment]
Some of this work had been done in this branch : https://github.com/spring-projects/spring-data-neo4j/tree/feature/reactive-support.
Don't know however which one is the more complete. @LilitGabrielyan it would be interesting to rebase against this branch. 

It was decided to leave it as is for the moment, as work is undergoing on OGM and neo4j drivers to have a real full stack reactive support.

**Comment by LilitGabrielyan** [comment]
Thanks Nicolas for reaching out.
I have seen that some work was done in that branch but that wasn't enough,
so I decided to write from scratch. I can try to rebase the branch in order
to have one complete branch.
Let me know what you prefer and should I rebase from my forked repository
branch.

Thanks,
Lilit

On Mon, Nov 6, 2017 at 1:44 PM, Nicolas Mervaillie <notifications@github.com

**Comment by michael-simons** [comment]
Hi @LilitGabrielyan. Thanks a lot for that massive PR.

As it stand today, there's a lot more work involved that just rebasing your PR.
Currently we are missing a reactive driver (Java Driver is async, but doesn't support back pressure etc.) and also the necessary infrastructure in OGM itself.

I fully agree with everything Mark commented already. The same is true for Nicolas' branch.

I therefor close this PR.

**Pull request closed without merging**